library/qtest/src/Operations.qs

```diff
@@ -0,0 +1,116 @@
+/// if all tests passed, and `false` if any one test failed.
+/// Prints the expected and received result for any failed tests.
+///
+/// # Input
+/// ## test_cases
+/// An array of five-arity tuples of the form `(test_name, num_qubits, qubit_prep_callable, callable_to_test, expected_result)`.
+/// `num_qubits` will be allocated in a qubit array and passed to `qubit_prep_callable` to prepare the state before executing 
+/// `callable_to_test`. Afterwards, `callable_to_test` will be called and its result will be compared to `expected_result`.
+///
+/// # Example
+/// ```qsharp
+/// function Main() : Unit {
+///     TestCases([
+///         ("Should return 42", TestCaseOne, 42),
+///         ("Should add one", () => AddOne(5), 6)
+///     ]);
+/// }
+///
+/// function TestCaseOne() : Int {
+///     42
+/// }
+///
+/// function AddOne(x: Int) : Int {
+///     x + 1
+/// }
+/// ```
+operation TestCases<'Result : Eq + Show > (test_cases : (String, Int, (Qubit[]) => (), (Qubit[]) => 'Result, 'Result)[]) : Bool {
+    let failed_test_buf = TestCasesSilent(test_cases);
+    
+    if Length(failed_test_buf) == 0 {
+        Message($"{Length(test_cases)} test(s) passed.");
+        true
+    } else {
+        Message($"{Length(failed_test_buf)} tests failed.");
+        for item in failed_test_buf {
+            Message($"{item}")
+        }
+        false
+    }
+}
+
+/// # Summary
+/// Similar to `Qtest.Operations.TestCases`, but returns test failure info in an array of strings instead of directly messaging to
+/// output. Useful if you want to handle your own messaging when writing tests, for CI or similar.
+
+/// See `Qtest.Operations.TestCases` for more details.
+///
+/// # Description
+/// Given an array of test names, test functions, and expected results (of the same type), returns an array
+/// of strings representing all failed test cases (if any). Strings are of the form "test_name: expected {}, got {}"
+///
+/// # Input
+/// ## test_cases
+/// An array of five-arity tuples of the form `(test_name, num_qubits, qubit_prep_callable, callable_to_test, expected_result)`.
+/// `num_qubits` will be allocated in a qubit array and passed to `qubit_prep_callable` to prepare the state before executing 
+/// `callable_to_test`. Afterwards, `callable_to_test` will be called and its result will be compared to `expected_result`.
+///
+/// # Example
+/// ```qsharp
+/// function Main() : Unit {
+///     let failure_messages = TestCasesSilent([
+///         ("Should return 42", TestCaseOne, 42),
+///         ("Should add one", () => AddOne(5), 6)
+///     ]);
+///     Std.Diagnostics.Fact(Length(failure_messages) == 0, "No tests should fail.")
+/// }
+///
+/// function TestCaseOne() : Int {
+///     42
+/// }
+///
+/// function AddOne(x: Int) : Int {
+///     x + 1
+/// }
+/// ```
+operation TestCasesSilent<'Result : Eq + Show > (test_cases : (String, Int, (Qubit[]) => (), (Qubit[]) => 'Result, 'Result)[]) : String[] {
```

Comment:
How about naming this `TestCasesVerbose` and returning an array of (Bool, String). This would allow people to gather test execution feedback and do some sort of post processing.

Reply:
A good follow up might be to provide collective statistics: number of tests passed/failed along with which test case passed/failed. This might be really helpful in CI/CD. Here are some ambitious ideas:
- **VSCode:** How about a VSCode Cmd Pallete that has a GUI feel. For example: Python tests, Rust tests, etc.
- **CI/CD:** Possibly having an artifact for test case results that some one can open in VSCode This might require Rust interfacing since Q# doesnt support file I/O. And Q# extension could then show results in GUI form from test artifact
- **Python:** Having a `qsharp_test` library that does some convenient wraparound to eliminate wrappers . `samples/testing` has examples of Python interop.

Reply:
These are great ideas, @Manvi-Agrawal  -- thank you! We were just looking at integrating with VS Code first -- something using this perhaps https://code.visualstudio.com/api/extension-guides/testing

For CI/CD, we could support a standardized output format from Q# and parse that in Python, then dump out JSON for analysis. Something like that could be automated. What do you think about that?

Reply:
I'm resolving this so the PR is unblocked, but feel free to continue chatting in this thread, or file an issue and tag me to discuss further.

Reply:
Sounds good @sezna . Created #2029 to track. Feel free to modify issue description.